languages/python/sqlalchemy-oso/tests/test_post_relationship.py

```diff
@@ -0,0 +1,530 @@
+        'post.created_by.username = actor.username and post.access_level = "private";'
+    )
+
+    oso.load_str(
+        'allow(actor: User, "read", post: Post) if post.created_by.is_banned = false and '
+        'post.access_level = "public";'
+    )
+
+    # moderator can see posts made by banned users.
+    oso.load_str(
+        'allow(actor: User, "read", post: Post) if actor.is_moderator = true and post.created_by.is_banned = true;'
+    )
+
+    foo = session.query(User).filter(User.username == "foo").first()
+
+    posts = authorize_model(oso, foo, "read", session, Post)
+
+    def allowed(post, user):
+        return (
+            (post.access_level == "public" and post.created_by.is_banned == False)
+            or post.access_level == "private"
+            and post.created_by == user
+        )
+
+    assert posts.count() == 7
+    assert all(allowed(post, foo) for post in posts)
+
+    admin = session.query(User).filter(User.username == "admin_user").first()
+    posts = authorize_model(oso, admin, "read", session, Post)
+
+    def allowed_admin(post):
+        return post.created_by.is_banned
+
+    assert posts.count() == 6
+    for post in posts:
+        assert allowed(post, admin) or allowed_admin(post)
+
+
+@pytest.fixture
+def tag_test_fixture(session):
+    """Test data for tests with tags."""
+    user = User(username="user")
+    other_user = User(username="other_user")
+    moderator = User(username="moderator", is_moderator=True)
+
+    eng = Tag(name="eng")
+    random = Tag(name="random", is_public=True)
+
+    user_public_post = Post(
+        contents="public post", created_by=user, access_level="public"
+    )
+    user_private_post = Post(
+        contents="private user post", created_by=user, access_level="private"
+    )
+
+    other_user_public_post = Post(
+        contents="other user public", created_by=other_user, access_level="public"
+    )
+    other_user_private_post = Post(
+        contents="other user private", created_by=other_user, access_level="private"
+    )
+    other_user_random_post = Post(
+        contents="other user random",
+        created_by=other_user,
+        access_level="private",
+        tags=[random],
+    )
+
+    # HACK!
+    objects = {}
+    for (name, local) in locals().items():
+        if name != "session" and name != "objects":
+            session.add(local)
+
+        objects[name] = local
+
+    session.commit()
+
+    return objects
+
+
+def test_in_multiple_attribute_relationship(session, oso, tag_test_fixture):
+    oso.load_str(
+        """
+        allow(user, "read", post: Post) if post.access_level = "public";
+        allow(user, "read", post: Post) if post.access_level = "private" and post.created_by = user;
+        allow(user, "read", post: Post) if tag in post.tags and tag.is_public = true;
+    """
+    )
+
+    posts = authorize_model(oso, tag_test_fixture["user"], "read", session, Post)
+    print(str(posts.statement.compile()))
```

Comment:
TODO: remove prints?

Reply:
🤷  they are helpful if something fails, can remove if you want